prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1180 @@
+		mrecv_entry->multi_recv_ref++;
+		rx_entry->peer_entry.owner_context = mrecv_entry;
+
+		if (util_adjust_multi_recv(srx, &mrecv_entry->peer_entry,
+					   rx_entry->peer_entry.size))
+			buf_done = true;
+
+		srx->update_func(srx, rx_entry);
+		ret = rx_entry->peer_entry.srx->peer_ops->start_msg(
+							&rx_entry->peer_entry);
+		if (ret || buf_done)
+			goto out;
+
+		rx_entry = util_search_unexp_msg(srx, addr);
+	}
+
+	queue = addr == FI_ADDR_UNSPEC ? &srx->msg_queue:
+		ofi_array_at(&srx->src_recv_queues, addr);
+	assert(queue);
+	slist_insert_tail((struct slist_entry *)(&mrecv_entry->peer_entry),
+			  queue);
+out:
+	ofi_genlock_unlock(&srx->lock);
+	return ret;
+}
+
+static struct util_rx_entry *util_search_peer_tag(struct util_unexp_peer *peer,
+				uint64_t tag, uint64_t ignore, bool remove)
+{
+	struct util_rx_entry *rx_entry;
+	struct slist_entry *item, *prev;
+
+	assert(peer);
+	if (slist_empty(&peer->tag_queue))
+		return NULL;
+
+	slist_foreach(&peer->tag_queue, item, prev) {
+		rx_entry = (struct util_rx_entry *) item;
+		if (!ofi_match_tag(tag, ignore, rx_entry->peer_entry.tag))
+			continue;
+
+		if (remove) {
+			slist_remove(&peer->tag_queue, item, prev);
+			if (!--peer->cnt) {
+				assert(slist_empty(&peer->tag_queue));
+				dlist_remove(&peer->entry);
+			}
+		}
+		return rx_entry;
+	}
+	return NULL;
+}
+
+static struct util_rx_entry *util_search_unexp_tag(struct util_srx_ctx *srx,
+		fi_addr_t addr, uint64_t tag, uint64_t ignore, bool remove)
+{
+	struct util_rx_entry *rx_entry;
+	struct util_unexp_peer *unexp_peer;
+	struct dlist_entry *entry;
+
+	if (addr == FI_ADDR_UNSPEC) {
+		dlist_foreach(&srx->unspec_unexp_tag_queue, entry) {
+			rx_entry = container_of(entry, struct util_rx_entry,
+						peer_entry);
+			if (!ofi_match_tag(tag, ignore,
+			    rx_entry->peer_entry.tag))
+				continue;
+
+			if (remove)
+				dlist_remove(entry);
+
+			return rx_entry;
+		}
+
+		dlist_foreach_container(&srx->unexp_peers,
+				struct util_unexp_peer, unexp_peer, entry) {
+			rx_entry = util_search_peer_tag(unexp_peer, tag,
+							ignore, remove);
+			if (rx_entry)
+				return rx_entry;
+		}
+		return NULL;
+	}
+
+	return util_search_peer_tag(ofi_array_at(&srx->src_unexp_peers, addr),
+				    tag, ignore, remove);
+}
+
+static ssize_t util_srx_peek(struct util_srx_ctx *srx, const struct iovec *iov,
+			     void **desc, size_t iov_count, fi_addr_t addr,
+			     void *context, uint64_t tag, uint64_t ignore,
+			     uint64_t flags)
+{
+	struct util_rx_entry *rx_entry;
+	int ret = FI_SUCCESS;
+
+	rx_entry = util_search_unexp_tag(srx, addr, tag, ignore,
+					 flags & (FI_CLAIM | FI_DISCARD));
+	if (!rx_entry) {
+		FI_DBG(&core_prov, FI_LOG_EP_CTRL, "Message not found\n");
+		return ofi_cq_write_error_peek(srx->cq, tag, context);
+	}
+	FI_DBG(&core_prov, FI_LOG_EP_CTRL, "Message found\n");
+
+	if (flags & (FI_CLAIM | FI_DISCARD)) {
+		if (flags & FI_DISCARD) {
+			ret = rx_entry->peer_entry.srx->peer_ops->discard_tag(
```

Comment:
So we do not release rx entry in the discard op, but release it outside explicitly?

Reply:
I see, the man page has stated that discard call should only release resources associated with the fi_peer_rx_entry.